### ISSUE TITLE
Make ServiceBus clients and objects testeable for the SDK users

### ISSUE THREAD

**Issue by kelps** [Azure/azure-sdk-for-net#19758]
As of this moment, the main client, services and objects in **Azure.Messaging.ServiceBus** (track2) are not easily mockable, making it very difficult to unit test applications that use the SDK. 

Making a SDK like this, that depends heavily on cloud environment to run and not being able to mock it to unit test our code is a bad design decision, specially since it would've been very easy to simply define interfaces for the clients, services and objects to make it possible.

I am starting new developments now and I used **Microsoft.Azure.ServiceBus** (track1) in my architecture tests, but was planning on switching to the new SDK when I saw it is the new way forward. This track1 SDK is barely mockable (I only had to do a little Reflection in the Message object to set a private property), but worked. The track2 SDK clients, services and objects don't implement any interfaces that I could be mocking in my code, and the actual classes have LOTS of private inaccessible code, making it impractical or sometimes impossible to do the mocking by inheriting them.

The way I see, there are 2 possible paths to follow to "solve" this and improve things:

- option 1: define the required interfaces and change the classes to be implementations. To not break compatibility but allow for proper mock, the methods in the interfaces should expect only interfaces, structs or basic types and the current methods already in the classes should not change signature, but their implementation could move the the interface implementations [sample 1 bellow]
- option 2: make the current services, clients and objects properly inheritable, by turning the constructor, methods and properties overridable and create ways to tap into "private" code with "protected" methods when necessary (like when we need to raise an event).

**sample 1 - ServiceBusProcessor**
*current code snipet*
```csharp
public class ServiceBusProcessor: IAsyncDisposable{
  public event Func<ProcessMessageEventArgs, Task> ProcessMessageAsync;
  public event Func<ProcessErrorEventArgs, Task> ProcessErrorAsync;
  public virtual Task StartProcessingAsync(CancellationToken cancellationToken = default){...}
}
```

*proposed code snipet*
```csharp

public interface IProcessMessageEventArgs{
  IServiceBusReceivedMessage Message {get;}
  CancelationToken CancelationToken {get;}

  
}

public class ProcessMessageEventArgs : IProcessMessageEventArgs, EventArgs {
  IServiceBusReceivedMessage  IProcessMessageEventArgs.Message => this.Message;
  public ServiceBusReceivedMessage  Message {get;}
  public CancelationToken CancelationToken {get;}

  Task IProcessMessageEventArgs.CompleteMessageAsync(IServiceBusReceivedMessage message, CancellationToken cancellationToken = default){
    ...
  }

  public virtual Task CompleteMessageAsync(ServiceBusReceivedMessage message, CancellationToken cancellationToken = default) => CompleteMessageAsync(message, cancellationToken);

  ...
}

public interface IServiceBusProcessor: IAsyncDisposable{
  public event Func<IProcessMessageEventArgs, Task> ProcessMessageAsync;
  public event Func<IProcessErrorEventArgs, Task> ProcessErrorAsync;
  public virtual Task StartProcessingAsync(CancellationToken cancellationToken = default){...}
}
public class ServiceBusProcessor: IServiceBusProcessor{
  public virtual event Func<ProcessMessageEventArgs, Task> ProcessMessageAsync;
  public virtual event Func<ProcessErrorEventArgs, Task> ProcessErrorAsync;
  public virtual Task StartProcessingAsync(CancellationToken cancellationToken = default){...}
}
```

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by JoshLove-msft** [comment]
Hi @kelps,
What we are aiming for in this library, and all new Azure.* libraries, is essentially what you have laid out in Option 2. The gap that remains in Service Bus is that there is no way to override the event handlers for mocking. What we will likely do is add virtual `OnProcessMessageAsync` methods that can be overridden rather than making the events themselves virtual. This is

**Comment by kelps** [comment]
I though so. The event thing is one of my pain points, but not the only one. Many of the other clients, services and objects need to be completely inheritable and overrideable for a proper unit test to be possible.

Anyway, as long as the SDKs become more mockable, in any form, it is a win.

Are there any published roadmaps for the SDKs next releases? 

Thanks for the quick response.

**Comment by JoshLove-msft** [comment]
> Many of the other clients, services and objects need to be completely inheritable and overrideable for a proper unit test to be possible.

Would you be willing to provide some examples where they aren't?

> Are there any published roadmaps for the SDKs next releases?

Here is a [search](https://github.com/Azure/azure-sdk-for-net/issues?q=is%3Aopen+is%3Aissue+label%3A%22Service+Bus%22+milestone%3A%22%5B2021%5D+April%22) that shows the issues we are planning to include for our upcoming April beta release. I would like to try to include the OnProcessMessageAsync/OnProcessErrorAsync/etc methods in our next beta release for April, but it's possible it could get bumped to May.

**Comment by kelps** [comment]
@JoshLove-msft yes, absolutely! At the top of my head I can say that the ServiceBusMessage and ServiceBusReceivedMessage are good examples of classes hard to use in unit tests. Since they are explicitly required in some methods, we can't run away from them, and sme properties, methods and even constructors are private and many things aren't virtual, so we can't inherit from them either.

I didn't look at every possibility in the SDK, and also didn't look at other Azure.*, but I suppose many might have similar deficiencies. Tomorrow morning I'll take a better look at my code to provide other examples (it is late in Brazil now and I'm not on my computer at the moment)

I am partial to using interfaces for such things instead of making my class overrideable in that way. Using interfaces made it a LOT easier to mock the track1 ServiceBus SDK. I still had to use reflection for one little piece, also in the Message class, but everything else was simpler.

**Comment by JoshLove-msft** [comment]
> At the top of my head I can say that the ServiceBusMessage and ServiceBusReceivedMessage are good examples of classes hard to use in unit tests. Since they are explicitly required in some methods, we can't run away from them, and sme properties, methods and even constructors are private and many things aren't virtual, so we can't inherit from them either.

For the ServiceBusReceivedMessage, you can create it using the ServiceBusModelFactory [method](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ServiceBusModelFactory.cs#L22). For ServiceBusMessage, it can be created with the constructor. In general, for model types, we make sure that they can be created either with a constructor, or for output only types, via a model factory.

> I am partial to using interfaces for such things instead of making my class overrideable in that way. Using interfaces made it a LOT easier to mock the track1 ServiceBus SDK. I still had to use reflection for one little piece, also in the Message class, but everything else was simpler.

We tend to avoid using interfaces for public APIs in the new set of libraries as they don't version well, and prefer using abstract classes where they are needed.

**Comment by kelps** [comment]
Good to know that there are ways of doing it. Maybe what is missing is documentation or samples. Anyway, thanks.

I'll take a better look at my code tomorrow and give more feedback here.

Also, I liked the approach for APIs and model classes. I suppose abstract classes fill the same role as interfaces in these scenarios, and factories solve the other half of the problem. I'll look into it.

**Comment by kelps** [comment]
I just took a look at my code and the events in ServiceBusProcessor are now the only ones preventing me from unit testing message arrival. I did changed to use the factory to instanciate ServiceBusReceivedMessage (thanks).

For now, I'll do some local "live" tests and keep the unit tests commented until de protected OnProcess* methods are available. I saw that you already created a PR for that 😁🎉. #fingerscrossed that it'll be integrated into the april release 🤞

**Comment by JoshLove-msft** [comment]
@kelps the PR was merged in. Feel free to test it out using our dev feeds - https://github.com/Azure/azure-sdk-for-net/blob/master/CONTRIBUTING.md#nuget-package-dev-feed

**Comment by kelps** [comment]
Already! That is awesome! I'll do that. Thanks!

**Comment by JoshLove-msft** [comment]
Hi @kelps, just wanted to check in to see if you were able to test out the new methods.

**Comment by kelps** [comment]
Hi @JoshLove-msft, sorry for not replying sooner.

Yes, I did test and it worked 😁. Thank you very much for such a quick fix! Now I'm waiting for the release before creating my PR changing the SDK to the new one.

**Comment by JoshLove-msft** [comment]
There will be a beta version with these changes released next week.
